File: contrib/modelnet/crl12-sdk/typescript/src/desktop_context.ts
// Copyright (c) 2026 The BTX developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or https://opensource.org/license/mit/.
/** Desktop context is view/draft only. Never HTTP, never localhost execute. */

export const CONTEXT_TYPE = "btx.cognitiveReserve.v1_2";
export const VIEW_PURPOSES = ["INSPECT", "COMPARE"] as const;
export const DRAFT_PURPOSES = [
  "DRAFT",
  "DRAFT_RESERVE_ALLOCATION",
  "DRAFT_RESEARCH_COMMITMENT",
  "DRAFT_CAPABILITY_ACQUISITION",
  "DRAFT_PRODUCT_REFERRAL",
] as const;

const FORBIDDEN_KEYS = new Set([
  "access_token",
  "refresh_token",
  "id_token",
  "token",
  "private_key",
  "secret",
  "mnemonic",
  "wallet_rpc",
  "prompt",
  "private_prompt",
  "local_path",
  "cwd",
  "automatic_spend_atoms",
  "grand_total",
  "combined_aum_auc",
  "invented_aum",
  "invented_auc",
]);
const FORBIDDEN_SUBSTR = ["127.0.0.1", "localhost", "/capital/allocations/", "/execute", "BEGIN PRIVATE KEY"];

export class DesktopContextError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(code + ": " + message);
    this.name = "DesktopContextError";
    this.code = code;
  }
}

export type PlannedOp = {
  operation_id: string;
  method: "GET" | "POST";
  path?: Record<string, string>;
  body?: Record<string, unknown>;
};

export type DesktopPlan = {
  type: typeof CONTEXT_TYPE;
  purpose: string;
  operations: PlannedOp[];
  http: false;
  localhost: false;
  execute: false;
  view_or_draft: true;
};

function walk(value: unknown): void {
  if (value && typeof value === "object") {
    if (Array.isArray(value)) {
      for (const item of value) walk(item);
      return;
    }
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      const lk = k.toLowerCase();
      if ((FORBIDDEN_KEYS.has(lk) || lk.endsWith("_secret")) && v !== null && v !== undefined && v !== "") {
        throw new DesktopContextError("SECRET_INLINE", "context must not carry " + k);
      }
      if (typeof v === "string") {
        const low = v.toLowerCase();
        for (const needle of FORBIDDEN_SUBSTR) {
          if (low.includes(needle.toLowerCase())) {
            throw new DesktopContextError("LOCALHOST_EXECUTE_FORBIDDEN", needle);
          }
        }
      }
      walk(v);
    }
    return;
  }
  if (typeof value === "string") {
    const low = value.toLowerCase();
    for (const needle of FORBIDDEN_SUBSTR) {
      if (low.includes(needle.toLowerCase())) {
        throw new DesktopContextError("LOCALHOST_EXECUTE_FORBIDDEN", needle);
      }
    }
  }
}

export function applyDesktopContext(ctx: Record<string, unknown>): DesktopPlan {
  walk(ctx);
  if (ctx.type !== CONTEXT_TYPE) {
    throw new DesktopContextError("UNKNOWN_CONTEXT_TYPE", String(ctx.type));
  }
  const btx = ctx.btx;
  if (!btx || typeof btx !== "object" || Array.isArray(btx)) {
    throw new DesktopContextError("MISSING_BTX", "btx object required");
  }
  const b = btx as Record<string, unknown>;
  const purpose = String(b.purpose ?? "");
  const allowed = new Set<string>([...VIEW_PURPOSES, ...DRAFT_PURPOSES]);
  if (!allowed.has(purpose)) {
    throw new DesktopContextError("VIEW_DRAFT_ONLY", purpose || "missing purpose");
  }
  const ops: PlannedOp[] = [];
  const asset = b.asset_ref;
  const projection = b.projection_ref;
  if ((VIEW_PURPOSES as readonly string[]).includes(purpose)) {
    if (typeof asset === "string" && asset) {
      ops.push({ operation_id: "getInstitutionalAsset", method: "GET", path: { id: asset } });
    }
    if (typeof projection === "string" && projection) {
      ops.push({ operation_id: "getPortfolioProjection", method: "GET", path: { id: projection } });
    }
    if (purpose === "COMPARE") {
      ops.push({ operation_id: "listInstitutionalMetrics", method: "GET" });
    }
  } else {
    const body: Record<string, unknown> = {
      requested_action: purpose === "DRAFT" ? "DRAFT_RESERVE_ALLOCATION" : purpose,
      execute: false,
    };
    if (typeof projection === "string" && projection) body.source_projection_ref = projection;
    ops.push({ operation_id: "preparePortfolioInstruction", method: "POST", body });
  }
  for (const op of ops) {
    if (op.operation_id === "executeAllocation") {
      throw new DesktopContextError("VIEW_DRAFT_ONLY", op.operation_id);
    }
  }
  return {
    type: CONTEXT_TYPE,
    purpose,
    operations: ops,
    http: false,
    localhost: false,
    execute: false,
    view_or_draft: true,
  };
}

export { applyDesktopContext as apply_desktop_context };

File: contrib/modelnet/crl12-sdk/python/desktop_context.py
# Copyright (c) 2026 The BTX developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or https://opensource.org/license/mit/.
"""Desktop context is view/draft only. Never HTTP, never localhost execute."""

from __future__ import annotations

from typing import Any

CONTEXT_TYPE = "btx.cognitiveReserve.v1_2"
VIEW_PURPOSES = frozenset({"INSPECT", "COMPARE"})
DRAFT_PURPOSES = frozenset(
    {
        "DRAFT",
        "DRAFT_RESERVE_ALLOCATION",
        "DRAFT_RESEARCH_COMMITMENT",
        "DRAFT_CAPABILITY_ACQUISITION",
        "DRAFT_PRODUCT_REFERRAL",
    }
)
ALLOWED_PURPOSES = VIEW_PURPOSES | DRAFT_PURPOSES

_FORBIDDEN_KEYS = frozenset(
    {
        "access_token",
        "refresh_token",
        "id_token",
        "token",
        "private_key",
        "secret",
        "mnemonic",
        "wallet_rpc",
        "prompt",
        "private_prompt",
        "local_path",
        "cwd",
        "automatic_spend_atoms",
        "grand_total",
        "combined_aum_auc",
        "invented_aum",
        "invented_auc",
    }
)
_FORBIDDEN_SUBSTR = (
    "127.0.0.1",
    "localhost",
    "/capital/allocations/",
    "/execute",
    "BEGIN PRIVATE KEY",
)


class DesktopContextError(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(f"{code}: {message}")
        self.code = code


def _walk(value: Any) -> None:
    if isinstance(value, dict):
        for k, v in value.items():
            lk = str(k).lower()
            if lk in _FORBIDDEN_KEYS or lk.endswith("_secret"):
                if v not in (None, "", [], {}):
                    raise DesktopContextError("SECRET_INLINE", f"context must not carry {k}")
            if isinstance(v, str):
                low = v.lower()
                for needle in _FORBIDDEN_SUBSTR:
                    if needle.lower() in low:
                        raise DesktopContextError("LOCALHOST_EXECUTE_FORBIDDEN", needle)
            _walk(v)
    elif isinstance(value, list):
        for item in value:
            _walk(item)
    elif isinstance(value, str):
        low = value.lower()
        for needle in _FORBIDDEN_SUBSTR:
            if needle.lower() in low:
                raise DesktopContextError("LOCALHOST_EXECUTE_FORBIDDEN", needle)


def apply_desktop_context(ctx: dict[str, Any]) -> dict[str, Any]:
    """Map an application context to typed HCP view/draft operations. No HTTP."""
    if not isinstance(ctx, dict):
        raise DesktopContextError("INVALID_CONTEXT", "object required")
    _walk(ctx)
    if ctx.get("type") != CONTEXT_TYPE:
        raise DesktopContextError("UNKNOWN_CONTEXT_TYPE", str(ctx.get("type")))
    btx = ctx.get("btx")
    if not isinstance(btx, dict):
        raise DesktopContextError("MISSING_BTX", "btx object required")
    purpose = str(btx.get("purpose", ""))
    if purpose not in ALLOWED_PURPOSES:
        raise DesktopContextError("VIEW_DRAFT_ONLY", purpose or "missing purpose")
    ops: list[dict[str, Any]] = []
    asset = btx.get("asset_ref")
    projection = btx.get("projection_ref")
    if purpose in VIEW_PURPOSES:
        if isinstance(asset, str) and asset:
            ops.append(
                {
                    "operation_id": "getInstitutionalAsset",
                    "method": "GET",
                    "path": {"id": asset},
                }
            )
        if isinstance(projection, str) and projection:
            ops.append(
                {
                    "operation_id": "getPortfolioProjection",
                    "method": "GET",
                    "path": {"id": projection},
                }
            )
        if purpose == "COMPARE":
            ops.append({"operation_id": "listInstitutionalMetrics", "method": "GET"})
    else:
        body: dict[str, Any] = {
            "requested_action": purpose if purpose != "DRAFT" else "DRAFT_RESERVE_ALLOCATION",
            "execute": False,
        }
        if isinstance(projection, str) and projection:
            body["source_projection_ref"] = projection
        ops.append(
            {
                "operation_id": "preparePortfolioInstruction",
                "method": "POST",
                "body": body,
            }
        )
    for op in ops:
        oid = op["operation_id"]
        if oid == "executeAllocation" or "execute" in oid.lower():
            raise DesktopContextError("VIEW_DRAFT_ONLY", oid)
        path = str(op.get("path", "")) + str(op)
        if "/execute" in path or "127.0.0.1" in path:
            raise DesktopContextError("LOCALHOST_EXECUTE_FORBIDDEN", path)
    return {
        "type": CONTEXT_TYPE,
        "purpose": purpose,
        "operations": ops,
        "http": False,
        "localhost": False,
        "execute": False,
        "view_or_draft": True,
    }


__all__ = (
    "ALLOWED_PURPOSES",
    "CONTEXT_TYPE",
    "DRAFT_PURPOSES",
    "DesktopContextError",
    "VIEW_PURPOSES",
    "apply_desktop_context",
)
